src/connector/src/parser/chunk_builder.rs

```diff
@@ -14,83 +14,201 @@
 
 use std::sync::LazyLock;
 
+use risingwave_common::array::stream_record::RecordType;
 use risingwave_common::array::{ArrayBuilderImpl, Op, StreamChunk};
 use risingwave_common::bitmap::BitmapBuilder;
 use risingwave_common::log::LogSuppresser;
 use risingwave_common::types::{Datum, DatumCow, ScalarRefImpl};
 use risingwave_common::util::iter_util::ZipEqFast;
 use risingwave_connector_codec::decoder::{AccessError, AccessResult};
 use risingwave_pb::plan_common::additional_column::ColumnType as AdditionalColumnType;
+use smallvec::SmallVec;
 use thiserror_ext::AsReport;
 
 use super::MessageMeta;
 use crate::parser::utils::{
     extract_cdc_meta_column, extract_header_inner_from_meta, extract_headers_from_meta,
     extract_subject_from_meta, extract_timestamp_from_meta,
 };
-use crate::source::{SourceColumnDesc, SourceColumnType, SourceMeta};
+use crate::source::{SourceColumnDesc, SourceColumnType, SourceCtrlOpts, SourceMeta};
+
+/// Maximum number of rows in a transaction. If a transaction is larger than this, it will be force
+/// committed to avoid potential OOM.
+const MAX_TRANSACTION_SIZE: usize = 4096;
+
+/// Represents an ongoing transaction.
+struct Transaction {
+    id: Box<str>,
+    len: usize,
+}
 
 /// A builder for building a [`StreamChunk`] from [`SourceColumnDesc`].
 pub struct SourceStreamChunkBuilder {
-    descs: Vec<SourceColumnDesc>,
+    column_descs: Vec<SourceColumnDesc>,
+    source_ctrl_opts: SourceCtrlOpts,
     builders: Vec<ArrayBuilderImpl>,
     op_builder: Vec<Op>,
     vis_builder: BitmapBuilder,
+    ongoing_txn: Option<Transaction>,
+    ready_chunks: SmallVec<[StreamChunk; 1]>,
```

Comment:
`SmallVec` looks good.

Can you please elaborate in which (unusual) case we'll have multiple ready chunks stashed?

Reply:
When `chunk_size` is limited to `1`, and somehow we parsed 3 rows from one `SourceMessage`.

Reply:
Added a comment for the chunk builder in https://github.com/risingwavelabs/risingwave/pull/19863/commits/307ebc2eb6654115955110ee518fc882feb7d813